### PR TITLE
exec: support mixed type binary operators

### LINKED ISSUE #39189: exec: mixed type binary expressions and comparisons
We need to handle mixed types in projection ops. For example:

```
root@127.0.0.1:60323/defaultdb> create table foo (a int4 primary key);
CREATE TABLE

root@127.0.0.1:60323/defaultdb> insert into foo values(1);
INSERT 1

root@127.0.0.1:60323/defaultdb> select a+1 from foo;
   ?column?
+------------+
  2
(1 row)

root@127.0.0.1:60323/defaultdb> set experimental_vectorize=always;
SET

root@127.0.0.1:60323/?> select a+1 from foo;
pq: unable to vectorize execution plan: unable to columnarize render expression "@1 + 1:::INT8": BinaryExpr on int4 and int is unhandled
```

There are a few different types of mixed type expressions to cover. We should probably implement them in this order:
- [x] Comparison expressions. Only allow floats to be compared to to other float types, and ints to be compared to all other int types
- [x] Allow all the mixed type comparisons, like int-float, int-decimal, etc.
- [x] Binary expressions for plus, minus, and multiplication. Again, only mixed types should be float-float or int-int.
- [x] All the mixed type expressions for plus, minus, and mult. (Might be easier to do at same time as above?)
- [x] Binary expression for division. ~NOTE: currently the overloads do not implement the floor division `//` operator. start with that.~ (floor division work will be covered by https://github.com/cockroachdb/cockroach/issues/40312)
- [x] Then afterwards handle regular division with the `/` operator, which is not really done correctly right now.

Also, sometimes `CastExpr` is planned, and we will handle that in #39372

### PR DESCRIPTION
**Pull request by rafiss**
This adds support for all mixed type binary operators.

The type conversions are the same as the ones that happen in the
non-vectorized engine, in eval.go.

A few basic regression tests are added. Additional testing will be done
later using SQLSmith

closes #39189

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
Very nice work! I like that the framework you set up for this stuff is so easy to understand. I reviewed the first commit and left some comments.

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-:-LnKeyiXCC2zs95T5kDZ:b-jgkymu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson, @rafiss, and @yuzefovich)

___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKcKg567sNlJiBFXOQ:-LnKcKg567sNlJiBFXOR:b-px19wq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L100)):*
> ```Go
> 
> // anyTypeBinaryOpToOverloads maps a binary operator to all of the overloads
> // that implement it, including all mixed input types.
> ```

[nit]: it might be worth to explicitly call out that all overloads from `sameTypeBinaryOpToOverloads` are included here (same goes for the comparisons below).
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 201 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKd34A346tMrs4QNSQ:-LnKd34A346tMrs4QNSR:b-4fj40m) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L201)):*
> ```Go
> 				}
> 			}
> 			if leftType == rightType {
> ```

[nit]: I think it's cleaner to take out this `if` outside of the inner for loop so that `leftType` is well defined, but there is no `rightType`.
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 478 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKdcaKEjcfyCkA6Gol:-LnKdcaKEjcfyCkA6Gom:b-9q7y2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L478)):*
> ```Go
> 		// The float64 customizer is used to handle binOps with floats of different
> 		// widths, so we must cast to float64 in this case.
> 		if c.width == 64 {
> ```

[nit]: probably `s/is used to/used to/` (a few lines above).
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 521 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKdb_--aJXaA4c1v4u:-LnKdb_06TO4HBFiiiih:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L521)):*
> ```Go
> 		// The int64 customizer is used to handle binOps with integers of different
> 		// widths, so we must cast to int64 in this case.
> 		if c.width == 64 {
> ```

ditto
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 664 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKdxp11kA5GsvbOdRA:-LnKdxp11kA5GsvbOdRB:bsngadc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L664)):*
> ```Go
> 		t := template.Must(template.New("").Parse(`
> 			{
> 				tmpFloat, err := {{.Left}}.Float64()
> ```

I think, for correctness, we need to cast a float to a decimal and compare decimals.  When a decimal is outside of range for float64, an error will be emitted.
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 706 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKeI5HBBqZrSLCfCO5:-LnKeI5HBBqZrSLCfCO6:bu0zzg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L706)):*
> ```Go
> 				tmpDec := &apd.Decimal{}
> 				tmpDec.SetFinite(int64({{.Right}}), 0)
> 				if _, err := tree.DecimalCtx.{{.Op}}(&{{.Target}}, &{{.Left}}, tmpDec); err != nil {
> ```

I think Matt mentioned in another PR that `ExactCtx` should be used for comparing decimals.
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 874 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKeerH6lQW9bVAqTTN:-LnKeerH6lQW9bVAqTTO:b-hgda) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L874)):*
> ```Go
> func registerBinOpOutputTypes() {
> 	binOpOutputTypes = make(map[tree.BinaryOperator]map[coltypePair]coltypes.T)
> 	for binOp, _ := range binaryOpName {
> ```

[super nit]: if `binaryOpName[binOp]` is not used, might as well remove `, _` part completely.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40272)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson and @yuzefovich)

___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKcKg567sNlJiBFXOQ:-LnOJooK3ONVfxZuJHoI:bdcmr5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L100)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: it might be worth to explicitly call out that all overloads from `sameTypeBinaryOpToOverloads` are included here (same goes for the comparisons below).
</blockquote></details>

done, good idea.
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 201 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKd34A346tMrs4QNSQ:-LnOJqlh5tSh9odVzl5O:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L201)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: I think it's cleaner to take out this `if` outside of the inner for loop so that `leftType` is well defined, but there is no `rightType`.
</blockquote></details>

done
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 478 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKdcaKEjcfyCkA6Gol:-LnOKgvO8_Yg5iUm9_ud:b-rahkxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L478)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: probably `s/is used to/used to/` (a few lines above).
</blockquote></details>

i did mean "is used to" but i can reword this to be more clear.
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 521 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKdb_--aJXaA4c1v4u:-LnOLRlOB9PgYGHgzNVY:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L521)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto
</blockquote></details>

done
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 664 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKdxp11kA5GsvbOdRA:-LnOJJ9_7Ciyl0ZGAFX4:b-7lj14j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L664)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think, for correctness, we need to cast a float to a decimal and compare decimals.  When a decimal is outside of range for float64, an error will be emitted.
</blockquote></details>

I would have thought that too, but I noticed that in postgres, the result of `decimal <op> float` is always a float, so that's why I made this customizer work this way. However, I also realized that the CRDB typechecker does not allow `decimal <op> float` binary expressions so I guess we don't even need to support this anyway. It would be simpler for me to just remove this, I think.
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 706 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKeI5HBBqZrSLCfCO5:-LnOMadT07S5l56SaGXk:banhr9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L706)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think Matt mentioned in another PR that `ExactCtx` should be used for comparing decimals.
</blockquote></details>

hmm if so, maybe we should change that in a different PR. the existing vec code already uses `tree.DecimalCtx`. see `decimalCustomizer` above

```
func (decimalCustomizer) getBinOpAssignFunc() assignFunc {
	return func(op overload, target, l, r string) string {
		return fmt.Sprintf("if _, err := tree.DecimalCtx.%s(&%s, &%s, &%s); err != nil { execerror.NonVectorizedPanic(err) }",
			binaryOpDecMethod[op.BinOp], target, l, r)
	}
}
```
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 874 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKeerH6lQW9bVAqTTN:-LnOMq6R9qh7_SIoAah5:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L874)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[super nit]: if `binaryOpName[binOp]` is not used, might as well remove `, _` part completely.
</blockquote></details>

done

**Review by yuzefovich — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r2, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-:-LnOoi_n3KwALaZVDH1J:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mjibson, @rafiss, and @yuzefovich)

___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 478 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKdcaKEjcfyCkA6Gol:-LnOlESOC90FBZzMdzcd:b-l7mlml) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L478)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i did mean "is used to" but i can reword this to be more clear.
</blockquote></details>

I know that English is not my first language, but I was taught that there are two constructions with word "used" - one is like "I used to go to school", and another is like "I'm used to go*ing* to school", and the previous version of the comment didn't match either of those options, so that's why when I was reading it, I was like, something feels odd here.
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 664 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKdxp11kA5GsvbOdRA:-LnOm4-uCUBR59r9yu0g:bahiged) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L664)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

I would have thought that too, but I noticed that in postgres, the result of `decimal <op> float` is always a float, so that's why I made this customizer work this way. However, I also realized that the CRDB typechecker does not allow `decimal <op> float` binary expressions so I guess we don't even need to support this anyway. It would be simpler for me to just remove this, I think.
</blockquote></details>

Ok, this sounds good to me.
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 706 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKeI5HBBqZrSLCfCO5:-LnOm9uZ0XX50yKUNkzV:b-k6zw3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L706)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hmm if so, maybe we should change that in a different PR. the existing vec code already uses `tree.DecimalCtx`. see `decimalCustomizer` above

```
func (decimalCustomizer) getBinOpAssignFunc() assignFunc {
	return func(op overload, target, l, r string) string {
		return fmt.Sprintf("if _, err := tree.DecimalCtx.%s(&%s, &%s, &%s); err != nil { execerror.NonVectorizedPanic(err) }",
			binaryOpDecMethod[op.BinOp], target, l, r)
	}
}
```
</blockquote></details>

Yeah, I agree that it should not be a part of this PR. Please leave a TODO. I think the issue about the context is assigned to Jordan.
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 807 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnOmdtQDpD6ovXMtyk2:-LnOmdtQDpD6ovXMtyk3:b-495plk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L807)):*
> ```Go
> }
> 
> func (c floatIntCustomizer) getBinOpAssignFunc() assignFunc {
> ```

Type checker also doesn't allow float vs int comparison?
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 111 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnOkwtEDf2O3qYDsNbf:-LnOkwtF6CGjAPdxAS1X:bn4vp4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/09351dc83a1609436256d3f07795ef8e7fa4210b/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L111)):*
> ```Go
> // overloads that implement it, including all mixed type comparisons. IT also
> // includes all entries from sameTypeComparisonOpToOverloads.
> var anyTypeComparisonOpToOverloads map[tree.ComparisonOperator][]*overload
> ```

[nit]: `s/IT/It/` (above).

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40272)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mjibson and @yuzefovich)

___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 478 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKdcaKEjcfyCkA6Gol:-LnOrbmy72VYd0KdOu5P:b-q4a7jy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L478)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I know that English is not my first language, but I was taught that there are two constructions with word "used" - one is like "I used to go to school", and another is like "I'm used to go*ing* to school", and the previous version of the comment didn't match either of those options, so that's why when I was reading it, I was like, something feels odd here.
</blockquote></details>

Interesting! I actually enjoy discussing grammar, and learning what it's like to learn English as a non-native speaker, so I'm happy to explain more. Though I apologize if anything I say is already obvious to you or is boring. :) 

I completely understand your confusion. I was using was the passive voice, which is something that most arbiters of English style discourage, often because it can obscure the meaning of a sentence. https://en.wikipedia.org/wiki/English_passive_voice

A more verbose version of the original comment would be: "The float64 customizer is used in order to handle binOps with floats of different widths, so we must cast to float64 in this case."

When it's written like that, it's more clear that the sentence did not have the "used to" construction. Instead, it had the more common meaning of the word "use."
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 706 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKeI5HBBqZrSLCfCO5:-LnOqiFY7T2kikLvLXZz:bhnwk7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L706)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Yeah, I agree that it should not be a part of this PR. Please leave a TODO. I think the issue about the context is assigned to Jordan.
</blockquote></details>

added a todo
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 807 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnOmdtQDpD6ovXMtyk2:-LnOqm0vBgDGzCCw-Bfh:b-ygqkxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L807)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Type checker also doesn't allow float vs int comparison?
</blockquote></details>

it does, but that is handled by `func (c floatIntCustomizer) getCmpOpCompareFunc()`
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 111 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnOkwtEDf2O3qYDsNbf:-LnOrZek3huMQ0nsaHtP:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/09351dc83a1609436256d3f07795ef8e7fa4210b/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L111)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: `s/IT/It/` (above).
</blockquote></details>

done

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40272)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mjibson, @rafiss, and @yuzefovich)

___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 478 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnKdcaKEjcfyCkA6Gol:-LnOv9SX7MXj0kbIX2pX:b-23mu6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L478)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

Interesting! I actually enjoy discussing grammar, and learning what it's like to learn English as a non-native speaker, so I'm happy to explain more. Though I apologize if anything I say is already obvious to you or is boring. :) 

I completely understand your confusion. I was using was the passive voice, which is something that most arbiters of English style discourage, often because it can obscure the meaning of a sentence. https://en.wikipedia.org/wiki/English_passive_voice

A more verbose version of the original comment would be: "The float64 customizer is used in order to handle binOps with floats of different widths, so we must cast to float64 in this case."

When it's written like that, it's more clear that the sentence did not have the "used to" construction. Instead, it had the more common meaning of the word "use."
</blockquote></details>

Oh, yeah, I see it now :) For some reason, my mind was fixed on the special "used to" construction.
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 807 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40272#-LnOmdtQDpD6ovXMtyk2:-LnOvP1cEx5soFSzCsb0:b-xsvqgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1673833598bd53b9c19035a7954ef8265230954c/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L807)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it does, but that is handled by `func (c floatIntCustomizer) getCmpOpCompareFunc()`
</blockquote></details>

Yeah, I missed that, never mind.

**Comment by rafiss** [comment]
thanks so much for the great review, Yahor!

bors r+

**Comment by mjibson** [comment]
I should have commented earlier but

> Additional testing will be done later using SQLSmith

I recommend not relying on sqlsmith to find bugs. We should not be waiting for sqlsmith to find bugs and then add test cases only after the fact. sqlsmith finds things that we probably already should have had tests for but didn't. Writing good tests is hard, and sqlsmith is one tool to verify that our tests are sufficient. But when adding new features that are risky, we should actively be writing good tests. A good test is one that:

- covers the normal, intended, happy case usage
- covers the known edge cases
- covers the non-error branches at least once

Good tests need not be exhaustive, but they should at least test a large chunk of the known space.

sqlsmith makes up for our deficiencies where we were unable to come up with a test case that tested one of the above things. It can increase our confidence in large changes, new features, and refactors. However we don't depend on sqlsmith finding bugs because it might not ever run your code. Instead we should do good effort to write some solid test cases.

**Comment by rafiss** [comment]
@mjibson agreed - my PR description does not make it clear, but I do believe the tests I included in this PR cover the 3 basic points you mention. In fact, I would add one more and say that the tests we write should also cover the known error cases when possible to make sure they are handled correctly.

It is true, though, that there will be additional testing using sqlsmith, so that is why I mentioned that.

**Comment by mjibson** [comment]
:+1: glad to hear it!
